mldsa/poly.c

```diff
@@ -417,15 +417,37 @@ void poly_uniform(poly *a, const uint8_t seed[MLDSA_SEEDBYTES], uint16_t nonce)
  * Returns number of sampled coefficients. Can be smaller than len if not enough
  * random bytes were given.
  **************************************************/
+#if MLDSA_ETA == 2
+#define POLY_UNIFORM_ETA_NBLOCKS \
+  ((136 + STREAM256_BLOCKBYTES - 1) / STREAM256_BLOCKBYTES)
+#elif MLDSA_ETA == 4
+#define POLY_UNIFORM_ETA_NBLOCKS \
+  ((227 + STREAM256_BLOCKBYTES - 1) / STREAM256_BLOCKBYTES)
+#else
+#error "Invalid value of MLDSA_ETA"
+#endif
 static unsigned int rej_eta(int32_t *a, unsigned int len, const uint8_t *buf,
                             unsigned int buflen)
+__contract__(
+  requires(len <= buflen && len <= MLDSA_N && \
+                  buflen <= (POLY_UNIFORM_ETA_NBLOCKS * STREAM256_BLOCKBYTES))
+  requires(memory_no_alias(a, sizeof(int32_t) * len))
+  requires(memory_no_alias(buf, buflen))
+  assigns(memory_slice(a, sizeof(int32_t) * len))
+  ensures(return_value <= len)
+  ensures(array_abs_bound(a, 0, return_value, MLDSA_ETA + 1))
+)
 {
   unsigned int ctr, pos;
-  uint32_t t0, t1;
+  int32_t t0, t1;
```

Comment:
I don't think this is a good idea. `t0` and `t1` are conceptually unsigned values. 
Can you instead add an explicit cast to `int32_t` to the assignments to `a[...]`? That would be a cleaner solution.

Reply:
You need to cast `t0` and `t1` to `int32_t` before doing the subtraction, so `a[..] = 2 - (int32_t) t`. Otherwise, the subtraction is computed in `uint32_t`, and does indeed unverflow.

Reply:
If you define `t0,t1` as `int32_t`, CBMC would not catch an underflow in `t0 = t0 - (205 * t0 >> 10) * 5;`, which makes the proof weaker.